distributed/worker_state_machine.py

```diff
@@ -0,0 +1,342 @@
+        )
+
+
+class UniqueTaskHeap(Collection[TaskState]):
+    """A heap of TaskState objects ordered by TaskState.priority.
+    Ties are broken by string comparison of the key. Keys are guaranteed to be
+    unique. Iterating over this object returns the elements in priority order.
+    """
+
+    __slots__ = ("_known", "_heap")
+    _known: set[str]
+    _heap: list[tuple[tuple[int, ...], str, TaskState]]
+
+    def __init__(self):
+        self._known = set()
+        self._heap = []
+
+    def push(self, ts: TaskState) -> None:
+        """Add a new TaskState instance to the heap. If the key is already
+        known, no object is added.
+
+        Note: This does not update the priority / heap order in case priority
+        changes.
+        """
+        assert isinstance(ts, TaskState)
+        if ts.key not in self._known:
+            assert ts.priority
+            heapq.heappush(self._heap, (ts.priority, ts.key, ts))
+            self._known.add(ts.key)
+
+    def pop(self) -> TaskState:
+        """Pop the task with highest priority from the heap."""
+        _, key, ts = heapq.heappop(self._heap)
+        self._known.remove(key)
+        return ts
+
+    def peek(self) -> TaskState:
+        """Get the highest priority TaskState without removing it from the heap"""
+        return self._heap[0][2]
+
+    def __contains__(self, x: object) -> bool:
+        if isinstance(x, TaskState):
+            x = x.key
+        return x in self._known
+
+    def __iter__(self) -> Iterator[TaskState]:
+        return (ts for _, _, ts in sorted(self._heap))
+
+    def __len__(self) -> int:
+        return len(self._known)
+
+    def __repr__(self) -> str:
+        return f"<{type(self).__name__}: {len(self)} items>"
+
+
+class Instruction:
+    """Command from the worker state machine to the Worker, in response to an event"""
+
+    __slots__ = ()
+
+
+# TODO https://github.com/dask/distributed/issues/5736
+
+# @dataclass
+# class GatherDep(Instruction):
+#    __slots__ = ("worker", "to_gather")
+#    worker: str
+#    to_gather: set[str]
+
+
+# @dataclass
+# class FindMissing(Instruction):
+#    __slots__ = ()
+
+
+# @dataclass
+# class Execute(Instruction):
+#    __slots__ = ("key", "stimulus_id")
+#    key: str
+#    stimulus_id: str
+
+
+class SendMessageToScheduler(Instruction):
+    __slots__ = ()
+    #: Matches a key in Scheduler.stream_handlers
+    op: ClassVar[str]
+
+    def __init_subclass__(cls, op: str):
+        cls.op = op
+
+    def to_dict(self) -> dict[str, Any]:
+        """Convert object to dict so that it can be serialized with msgpack"""
+        d = {k: getattr(self, k) for k in self.__annotations__}
+        d["op"] = self.op
+        return d
+
+
+# Note: as of Python 3.10.2, @dataclass(slots=True) doesn't work with __init__subclass__
+# https://bugs.python.org/issue46970
+@dataclass
+class TaskFinishedMsg(SendMessageToScheduler, op="task-finished"):
```

Comment:
Is there a functional difference to the case where I simply define the op in our subclasses, e.g.

```python
class TaskFinishedMsg(SendMessageToScheduler):
    op = "task-finished"
```

The usage of metaclasses feels a bit complex. is there anything else going on that I'm not aware of or is this a style question?

Reply:
No, it's just style. Happy to remove it.